chore: add treefmt and format config.

tshaynik opened the pull request:
# Context

_Reason for the change? If an issue exists, reference it here using a [keyword](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)_

It is useful to have treefmt as a single command that can format many types of files across the repository.
#812 introduced this solution, but it is out of scope for that PR.

# Proposed Solution
Introduce a small PR to add a treefmt.toml file, provide a nix devshell with all necessary executables, and run the formatter on all relevant files.

# Important Changes Introduced
A nix devshell has been added with everything needed to run treefmt.

gytis-ivaskevicius commented:
I think changelogs are autogeneratd, we should tweak prettier config to follow their formatting

rhyslbw commented:
> I think changelogs are autogeneratd, we should tweak prettier config to follow their formatting

Yes, changelogs are generated with Lerna so best to ignore `CHANGELOG.md`

rhyslbw commented on packages/e2e/local-network/templates/babbage/byron-configuration.yaml:
```diff
@@ -120,7 +118,7 @@ defaultBackends:
 setupScribes:
   - scKind: FileSK
-    scName: "logs/testnet.log"
+    scName: 'logs/testnet.log'
```
This formatting is being inherited from an upstream source, so probably best to ignore anything in `packages/e2e/local-network/templates/`

rhyslbw reviewed (commented):
Is there a way we can trigger this to run as part of our existing `yarn lint` command, without nix or the need to install the binary manually?

gytis-ivaskevicius commented:
We wire these formatters through `treefmt` which is provided by `nix`, one may install nix/direnv to have full integration. Unfortunately as of yet there is no way to install `treefmt` via npm https://github.com/numtide/treefmt#installation

tshaynik commented:
Needs to be re-approved after excluding `packages/e2e/local-network/templates/`. Should  be ready to merge.

gytis-ivaskevicius reviewed (changes requested):
We need to use prettier from node_modules and make sure that tests pass (maybe rebase will fix that?)

tshaynik commented:
Just rebased, shifted to using the prettier from node_modules (tested and it worked), and added nodejs (and removed nixpkgs prettier) from the devshell.

tshaynik commented on treefmt.toml:
```diff
@@ -6,7 +6,7 @@ excludes = ["yarn-project.nix"]
 
 [formatter.prettier]
-command = "prettier"
+command = "node_modules/prettier/bin-prettier.js"
```
You're right, it works the current way, but your way is better.

tshaynik replied:
updated

blaggacao commented:
https://github.com/input-output-hk/cardano-js-sdk/pull/855 supersedes this adding git blame ignores for the reformatting commits. This helps reducing some blame noise (including in GitHub) and I hope it's going to be helpful down the line.

blaggacao reviewed (approved):
LGTM

blaggacao commented:
> It would be nice to have some instructions on common editors (vscode, webstorm) setup for the used formatters

Is barren: https://numtide.github.io/treefmt/integrations/

My Kakoune config: :smile: 
```
# set global formatter
hook global BufSetOption .* %{
    set-option buffer formatcmd "treefmt --stdin %val{buffile}"
}
hook global WinSetOption .* %{
  set window formatcmd "treefmt --stdin %val{buffile}"
}
hook global BufWritePre .* %{
    evaluate-commands %sh{
        if [ -n "$kak_opt_formatcmd" ]; then
            echo format
        fi
    }
}
```
